Improvements to the PWM Fan usermod

qistoph opened the pull request:
This is a combined PR for some improvements on the PWM Fan usermod.

**More precise RPM calculation**
Use milliseconds between actual measurements to calculate the desired speed instead of the configured seconds between checks.

**Show locked state when tacho is used**
The user interface shows **Speed**: `locked`/`auto` when the tacho pin is not used. This is useful information that is lost when the tacho pin is enabled, by showing **Speed**: `1500rpm`. This PR changes it to:
- **Speed**: `1500 rpm (locked)`
- **Speed**: `1500 rpm (auto)`

**Linear PWM control**
Before this PW the PWM speed is configured and calculated using
- Desired temperature (default: 35°C)
- Number of PWM steps (default: 7)
- Temperature difference per PWM step (default: 0.5°C)

<img width="320" height="320" alt="desmos-graph" src="https://github.com/user-attachments/assets/a80d1a1b-db72-41e6-8460-ebf16123109a" />

This PR changes the calculation to allow linear PWM control, instead of steps, and includes defaults to match the previous outcomes.
- Desired temperature (default: 34.5°C)
- Temperature at which fan should be at maximum speed (default: 38"C)

<img width="320" height="320" alt="desmos-graph(1)" src="https://github.com/user-attachments/assets/981ed993-85e3-4ddf-af35-3943a4c4ea1b" />

**Disable lock from UI with speed 0**
There was no way to disable the fan lock from the user interface. It would stay on until a reset or JSON API request. A user without access to the JSON API would not be able to turn to automatic fan control from the interface.

This PR changes the client side javascript call to unlock the automatic control when the user selects speed 0.

**Side effect**: the user can no longer disable the fan using the manual override, but would have to change the fan settings.


If any of these changes are undesired or require further improvement, please let me know. I'd be happy to help and look forward to have these changes included in the beautiful WLED.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Configurable maximum temperature ceiling to control fan max speed.
  * UI/state shows max-temperature control and explicit RPM labels for auto vs locked.

* **Bug Fixes**
  * More accurate RPM timing using measured intervals.
  * Temperature-to-PWM mapping changed from stepped to smooth linear interpolation with bounds/validation.
  * Manual PWM = 0 now clears any fan lock state.

* **Chores**
  * Persisted max-temperature setting across restarts; state updates reflect changes immediately.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

qistoph committed:
Fan: more precise RPM calculation

Use ms between actual measurements for calculation instead of configured seconds between checks

qistoph committed:
Fan: Show locked state when tacho is used

qistoph committed:
Fan: Lineair PWM control

qistoph committed:
Fan: Disable lock from UI with speed: 0

blazoncek reviewed (changes requested):
As an author and maintainer of the usermod I see no *real* added value in the proposed changes.
There is no benefit if fan is running at 33.333% of speed compared to 7-step 28.5% or 42.8%.
There are several other issues I could spot.

qistoph commented on usermods/PWM_fan/PWM_fan.cpp:
```diff
@@ -249,15 +248,16 @@ class PWMFanUsermod : public Usermod {
         uiDomString += F("':{'");
         uiDomString += FPSTR(_speed);
-        uiDomString += F("':parseInt(this.value)}});\" oninput=\"updateTrail(this);\" max=100 min=0 type=\"range\" value=");
+        uiDomString += F("':parseInt(this.value),'lock':parseInt(this.value)>0}});\" oninput=\"updateTrail(this);\" max=100 min=0 type=\"range\" value=");
```
"speed" is set to the int value.
"lock" is set to 'true' for values >0, otherwise 'false'
Effectively unlocking (auto mode) when setting to 0.
Before this change it was not possible to unlock after locken from the Config menu in the UI.

qistoph replied:
Not being able to return to auto mode from the UI seems like a missing feature.

I see three solutions each with its own associated impact:
1. Introduce a new button/option in the UI to switch back to auto mode - UI looks change
2. Change UI logic - Front-end usage changes slightly
3. Change back-end logic - JSON API usage changes.

Option 2 has the least impact.

Fitting option 3 could be putting the same logic in the back-end: `if (speed<=0) { lockFan=false }`. This would impact users who would want to use `{"speed": 0, "lock": true}` to lock the in fan _off_ state.

Looking forward to your preference and thoughts

blazoncek commented:
> becoming less of a disturbance.

Apparently you must be using this usermod with different intention than the one used when developing it. The idea is to provide ventilation to a controller enclosure, not living room ventilation.

Adding code to already crowded WLED which yields little real benefit is not really desired. However this is a (not default) usermod and some wiggle room may be exerted. I am willing to accept linear (or non linear for that matter) speed adjustment, but only that. No need to change polling interval beyond the ability for user to select one.

Perhaps @srg74 would also like to be informed.

qistoph committed:
Fan: fix code review issue

When pwmMaxValue <= pwmMinValue (due to misconfiguration), the function returns without updating the fan speed, leaving it at its previous state. This could cause overheating if the fan isn't already running at an appropriate speed.

qistoph committed:
Fan: fix code review issue

Integer division evaluates before multiplication, causing minPWMValuePct / 100 to be 0 for all values less than 100. For example, with minPWMValuePct = 50:
- Current: 50 / 100 * 255 = 0 * 255 = 0 ❌
- Correct: (50 * 255) / 100 = 127 ✓

qistoph commented on usermods/PWM_fan/PWM_fan.cpp:
```diff
@@ -167,25 +165,23 @@ class PWMFanUsermod : public Usermod {
+      int pwmMinValue = minPWMValuePct / 100 * _pwmMaxValue;
+      int pwmMaxValue = maxPWMValuePct / 100 * _pwmMaxValue;
+
+      if (pwmMaxValue <= pwmMinValue) return;
```
Fixed, pushing soon

qistoph committed:
Fan: fix code review issues

**Prevent potential division by zero in RPM calculation.**

The expression msInterval / 1000 uses integer division. If msInterval < 1000 (less than 1 second), the result is 0, causing division by zero when dividing last_rpm.

While the loop check at line 218 ensures at least tachoUpdateSec * 1000 ms have elapsed, defensive programming should handle edge cases.

**Use absolute maximum PWM for failsafe, not configured maximum.**

When pwmMaxValue <= pwmMinValue due to misconfiguration, the failsafe sets the fan to pwmMaxValue, which is the lower of the two values. This could result in insufficient cooling.

For a true failsafe, use _pwmMaxValue (255) to ensure maximum cooling.

qistoph committed:
Fan: fix code review issues

Add validation to prevent division by zero.

The linear interpolation at line 185 computes (temp - targetTemperature) / (maxSpeedTemperature - targetTemperature). If maxSpeedTemperature == targetTemperature, this results in division by zero.

qistoph committed:
Revert "Fan: Disable lock from UI with speed: 0"

This reverts commit 44fab4449ce87733f5683229260292377d74620a.

DedeHai commented:
these look like solid changes to me. I do not use the UM nor have hardware to test it but if this does not break any existing setups this has my green light

blazoncek reviewed (commented):
Original request and URS for PWM fan usermod came from @srg74 and if he approves the changes I'm ok with the update, if not, neither am I.

softhack007 commented:
@DedeHai should we merge this?

softhack007 committed:
Merge branch 'main' into fan_improv

qistoph committed:
Fan: fix code review issues

qistoph committed:
Merge branch 'main' into fan_improv
